p32/src/shapes.rs

```diff
@@ -0,0 +1,191 @@
+    const NAME: &'static str = "Triangle";
+
+    fn area(&self) -> f32 {
+        let x1 = self.p1.x - self.p2.x;
+        let y1 = self.p1.y - self.p2.y;
+        let x2 = self.p2.x - self.p3.x;
+        let y2 = self.p2.y - self.p3.y;
+        let x3 = self.p3.x - self.p1.x;
+        let y3 = self.p3.y - self.p1.y;
+        (x1 * (y2 - y3) + x2 * (y3 - y1) + x3 * (y1 - y2)).abs() / 2.0
+    }
+
+    fn perimeter(&self) -> f32 {
+        let x1 = self.p1.x - self.p2.x;
+        let y1 = self.p1.y - self.p2.y;
+        let x2 = self.p2.x - self.p3.x;
+        let y2 = self.p2.y - self.p3.y;
+        let x3 = self.p3.x - self.p1.x;
+        let y3 = self.p3.y - self.p1.y;
+        (x1 * x1 + y1 * y1).sqrt() + (x2 * x2 + y2 * y2).sqrt() + (x3 * x3 + y3 * y3).sqrt()
+    }
+
+    fn scale(&mut self, factor: f32) {
+        self.p1.scale(factor);
+        self.p2.scale(factor);
+        self.p3.scale(factor);
+    }
+}
+
+impl Shape for Circle {
+    const NAME: &'static str = "Circle";
+
+    fn area(&self) -> f32 {
+        std::f32::consts::PI * self.radius * self.radius
+    }
+
+    fn perimeter(&self) -> f32 {
+        2.0 * std::f32::consts::PI * self.radius
+    }
+
+    fn scale(&mut self, factor: f32) {
+        self.center.scale(factor);
+        self.radius *= factor;
+    }
+}
+
+impl Shape for Rectangle {
+    const NAME: &'static str = "Rectangle";
+
+    fn area(&self) -> f32 {
+        ((self.p1.x - self.p2.x) * (self.p1.y - self.p2.y)).abs()
+    }
+
+    fn perimeter(&self) -> f32 {
+        2.0 * ((self.p1.x - self.p2.x).abs() + (self.p1.y - self.p2.y).abs())
+    }
+
+    fn scale(&mut self, factor: f32) {
+        self.p1.scale(factor);
+        self.p2.scale(factor);
+    }
+}
+
+impl Shape for DynamicShape {
+    const NAME: &'static str = "DynamicShape";
+
+    fn area(&self) -> f32 {
+        match self {
+            DynamicShape::Point(point) => point.area(),
+            DynamicShape::Circle(circle) => circle.area(),
+            DynamicShape::Triangle(triangle) => triangle.area(),
+            DynamicShape::Rectangle(rectangle) => rectangle.area(),
+        }
+    }
+
+    fn perimeter(&self) -> f32 {
+        match self {
+            DynamicShape::Point(point) => point.perimeter(),
+            DynamicShape::Circle(circle) => circle.perimeter(),
+            DynamicShape::Triangle(triangle) => triangle.perimeter(),
+            DynamicShape::Rectangle(rectangle) => rectangle.perimeter(),
+        }
+    }
+
+    fn scale(&mut self, factor: f32) {
+        match self {
+            DynamicShape::Point(point) => point.scale(factor),
+            DynamicShape::Triangle(triangle) => triangle.scale(factor),
+            DynamicShape::Circle(circle) => circle.scale(factor),
+            DynamicShape::Rectangle(rectangle) => rectangle.scale(factor),
+        };
+    }
+}
+
+pub enum Return {
+    First,
+    Second,
+}
+
+pub fn bigger_area_to_perimeter<'a>(first: &'a impl Shape, second: &'a impl Shape) -> Return {
```

Comment:
This function should accept two shape _slices_  and return _reference_ to the biggest shape.

Reply:
How should I use enum to return reference to a trait? Should I use DynamicShape wrapper?

Reply:
I will come back to this one later.

Reply:
You should make the enum itself generic. Each variant should contain a reference to a shape of different type.

Reply:
Got it. And should each slice have multiple shapes in them? And should I return the shape with the biggest area-to-perimeter among all the shapes in the two slices? Or just the slice with the larger total-area-to-total-perimeter?

Reply:
Each slice should contain shapes of one type. The return result should contain reference to one shape with the biggest ratio.

Reply:
Can you check now?